[CI] Various tests in ShrinkIndexIT fail with "expected at least one master-eligible node left"

This failure has occurred many times in the last 36 hours: https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:relative,to:now))&_a=(columns:!(test,build-id,build_url,branch),filters:!(),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'class:%20%22org.elasticsearch.action.admin.indices.create.ShrinkIndexIT%22'),sort:!(time,desc))

Often the failing test is `ShrinkIndexIT.testCreateShrinkWithIndexSort`, see for example https://gradle.com/s/fglguuelmqobe

More recently a failure of `ShrinkIndexIT.testShrinkCommitsMergeOnIdle` was seen but with the same error message: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/2405/testReport/junit/org.elasticsearch.action.admin.indices.create/ShrinkIndexIT/testShrinkCommitsMergeOnIdle/

In both cases the error message is something like:

```
java.lang.AssertionError: expected at least one master-eligible node left in {node_sc1=org.elasticsearch.test.InternalTestCluster$NodeAndClient@2de106c0}
```

A REPRO command for 7.x is:

```
./gradlew :server:integTest --tests "org.elasticsearch.action.admin.indices.create.ShrinkIndexIT.testCreateShrinkWithIndexSort" -Dtests.seed=50AF68DFDA0BACBE -Dtests.security.manager=true -Dtests.locale=zh-TW -Dtests.timezone=America/North_Dakota/Center -Dcompiler.java=12 -Druntime.java=8
```

A REPRO command for master is:

```
./gradlew :server:integTest --tests "org.elasticsearch.action.admin.indices.create.ShrinkIndexIT.testShrinkCommitsMergeOnIdle" -Dtests.seed=C57BD4CF13C2F399 -Dtests.security.manager=true -Dtests.locale=fo -Dtests.timezone=Europe/Zagreb -Dcompiler.java=12 -Druntime.java=11
```

Neither of these reproduce locally for me.

I will mute the suite.

--- a/server/src/test/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java
+++ b/server/src/test/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java
@@ -75,7 +75,6 @@ import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 
-@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
 public class ShrinkIndexIT extends ESIntegTestCase {
 
     @Override
@@ -559,7 +558,8 @@ public class ShrinkIndexIT extends ESIntegTestCase {
     }
 
     public void testShrinkThenSplitWithFailedNode() throws Exception {
-        internalCluster().ensureAtLeastNumDataNodes(3);
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        final String shrinkNode = internalCluster().startDataOnlyNode();
 
         final int shardCount = between(2, 5);
         prepareCreate("original").setSettings(Settings.builder().put(indexSettings())
@@ -567,8 +567,6 @@ public class ShrinkIndexIT extends ESIntegTestCase {
             .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, shardCount)).get();
         client().admin().indices().prepareFlush("original").get();
         ensureGreen();
-        final String shrinkNode
-            = client().admin().cluster().prepareNodesInfo("data:true").clear().get().getNodes().get(0).getNode().getName();
         client().admin().indices().prepareUpdateSettings("original")
             .setSettings(Settings.builder()
                 .put(IndexMetaData.INDEX_ROUTING_REQUIRE_GROUP_SETTING.getConcreteSettingForNamespace("_name").getKey(), shrinkNode)

--- a/server/src/test/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java
+++ b/server/src/test/java/org/elasticsearch/action/admin/indices/create/ShrinkIndexIT.java
@@ -75,7 +75,7 @@ import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 
-@ESIntegTestCase.AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/44164")
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
 public class ShrinkIndexIT extends ESIntegTestCase {
 
     @Override

--- a/test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java
@@ -1485,6 +1485,13 @@ public final class InternalTestCluster extends TestCluster {
         ensureOpen();
         NodeAndClient nodeAndClient = getRandomNodeAndClient(nc -> filter.test(nc.node.settings()));
         if (nodeAndClient != null) {
+            if (nodeAndClient.nodeAndClientId() < sharedNodesSeeds.length && nodeAndClient.isMasterEligible() && autoManageMasterNodes
+                && nodes.values().stream()
+                        .filter(NodeAndClient::isMasterEligible)
+                        .filter(n -> n.nodeAndClientId() < sharedNodesSeeds.length)
+                        .count() == 1) {
+                throw new AssertionError("Tried to stop the only master eligible shared node");
+            }
             logger.info("Closing filtered random node [{}] ", nodeAndClient.name);
             stopNodesAndClient(nodeAndClient);
         }

